File: Kit/Core/Components/Focus/BdkFocusFirstFieldOnInvalidSubmit.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace BlazorDevKit;

public partial class BdkFocusFirstFieldOnInvalidSubmit : ComponentBase, IAsyncDisposable
{
    private ElementReference _elementReference;
    private IJSObjectReference? _jsModuleReference;
    private IJSObjectReference? _jsInstanceReference;
    private readonly string _id = Guid.NewGuid().ToString("N");

    [Inject] public required IJSRuntime JsRuntime { get; set; }
    
    /// <summary>
    /// By default, the ‘.invalid, .is-invalid’ selector is added to the field that is invalid when the form is submitted and validated by EditContext.
    /// If you want to use a different class, you can set this parameter.
    /// </summary>
    [Parameter] public required string InvalidSelector { get; set; } = ".invalid, .is-invalid";

    /// <summary>
    /// Force the focus to the first invalid field if the element is not focusable.
    /// </summary>
    [Parameter] public bool Force { get; set; } = true;

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if(firstRender)
        {
            _jsModuleReference = await JsRuntime.InvokeAsync<IJSObjectReference>("import", "./_content/BlazorDevKit.Core/Components/Focus/BdkFocusFirstFieldOnInvalidSubmit.razor.js");
            _jsInstanceReference = await _jsModuleReference
                .InvokeAsync<IJSObjectReference>(
                    "BdkFocusFirstFieldOnInvalidSubmit.create",
                    _id,
                    DotNetObjectReference.Create(this),
                    _elementReference,
                    InvalidSelector
                );
            
        }
    }

    [JSInvokable]
    public async Task NotifySubmitAsync()
    {
        if(_jsInstanceReference is null) return;
        await _jsInstanceReference.InvokeVoidAsync("validationChangedAsync", Force);
    }
    
    public async ValueTask DisposeAsync()
    {
        try
        {
            if (_jsInstanceReference is not null)
            {
                await _jsInstanceReference.InvokeVoidAsync("dispose", _id);
                await _jsInstanceReference.DisposeAsync();
            }

            if (_jsModuleReference is not null)
            {
                await _jsModuleReference.DisposeAsync();
            }
        }
        catch (JSDisconnectedException)
        {
            // Ignore, see:
            // https://stackoverflow.com/questions/72488563/blazor-server-side-application-throwing-system-invalidoperationexception-javas
            // https://github.com/dotnet/aspnetcore/issues/49376
            // https://github.com/dotnet/aspnetcore/issues/30344
        }
    }
}

File: Kit/Core/Components/Focus/BdkFocusFirstFieldOnInvalidSubmit.razor.js
export class BdkFocusFirstFieldOnInvalidSubmit {

    constructor(dotNetReference, elementReference, invalidSelector) {
        this.invalidSelector = invalidSelector;
        this.dotNetReference = dotNetReference;
        this.form = elementReference.closest('form');
        this.disposed = false;
    }

    async validationChangedAsync(forceFocus) {
        if (this.disposed) { return; }
        let invalidInput = this.form.querySelector(this.invalidSelector);
        let attempts = 1;
        let maxAttempts = 4;        
        
        while (invalidInput == null && attempts <= maxAttempts) {
            await new Promise(resolve => setTimeout(resolve, 50));
            if (this.disposed) { return; }
            invalidInput = this.form.querySelector(this.invalidSelector);
            attempts++;
        }
        
        if(!invalidInput) { return; }
        
        if (this.isFocusable(invalidInput)) {
            invalidInput.focus();
        } else if (forceFocus) {
            invalidInput.tabIndex = 0;
            invalidInput.focus();
            invalidInput.tabIndex = -1;
        } else {
            console.log('The .invalid input is not focusable');
        }
    }    
    
    isFocusable(element) {
        return element.tabIndex >= 0;
    }
    
    dispose(identifier) {
        this.disposed = true;
        this.form = null;
        delete window[identifier];
    }

    static create(identifier, dotNetReference, elementReference, invalidSelector) {
        window[identifier] = new BdkFocusFirstFieldOnInvalidSubmit(dotNetReference, elementReference, invalidSelector);
        window[identifier].form.addEventListener('submit', () => {
            // Sometimes it may happen that the component is already disposed when the submit event is fired
            if (!window[identifier]) { return; }
            window[identifier].dotNetReference.invokeMethodAsync('NotifySubmitAsync');
        });
        return window[identifier];        
    }
}

window.BdkFocusFirstFieldOnInvalidSubmit = BdkFocusFirstFieldOnInvalidSubmit;